src/libos/src/blk/dev_disk.rs

```diff
@@ -0,0 +1,182 @@
+                        RawDisk::open_or_create(total_blocks, image_path.to_str().unwrap())?;
+                    let root_key = metadata.root_key;
+
+                    let sworndisk = Arc::new(
+                        SwornDisk::open(raw_disk.clone(), root_key, None).unwrap_or_else(|_e| {
+                            SwornDisk::create(raw_disk, root_key, None).unwrap()
+                        }),
+                    );
+                    sworndisk_opt.insert(sworndisk.clone());
+                    sworndisk
+                }
+            }
+            _ => return_errno!(EINVAL, "Unrecognized block device"),
+        };
+        Ok(Self { disk })
+    }
+
+    pub fn disk(&self) -> Arc<dyn BlockDevice> {
+        self.disk.clone()
+    }
+}
+
+// Used for registering the disk to the DevFS
+impl INode for DevDisk {
+    fn read_at(&self, offset: usize, buf: &mut [u8]) -> vfs::Result<usize> {
+        if rw_args_block_aligned(offset, buf.len()) {
+            self.disk
+                .read_blocks((offset / BLOCK_SIZE) as _, &mut [buf])?;
+        } else {
+            self.disk.read_bytes(offset, buf)?;
+        }
+
+        Ok(buf.len())
+    }
+
+    fn write_at(&self, offset: usize, buf: &[u8]) -> vfs::Result<usize> {
+        if rw_args_block_aligned(offset, buf.len()) {
+            self.disk.write_blocks((offset / BLOCK_SIZE) as _, &[buf])?;
+        } else {
+            self.disk.write_bytes(offset, buf)?;
+        }
+
+        Ok(buf.len())
+    }
+
+    fn metadata(&self) -> vfs::Result<Metadata> {
+        Ok(Metadata {
+            dev: 0,
+            inode: 0xfe23_1d08,
+            size: self.disk.total_bytes(),
+            blk_size: BLOCK_SIZE,
+            blocks: self.disk.total_blocks(),
+            atime: Timespec { sec: 0, nsec: 0 },
+            mtime: Timespec { sec: 0, nsec: 0 },
+            ctime: Timespec { sec: 0, nsec: 0 },
+            type_: FileType::File,
+            mode: 0o666,
+            nlinks: 1,
+            uid: 0,
+            gid: 0,
+            rdev: 0,
+        })
+    }
+
+    fn sync_all(&self) -> vfs::Result<()> {
+        self.disk.sync()?;
+        Ok(())
+    }
+
+    fn sync_data(&self) -> vfs::Result<()> {
+        self.disk.sync()?;
+        Ok(())
+    }
+
+    fn as_any_ref(&self) -> &dyn Any {
+        self
+    }
+}
+
+fn rw_args_block_aligned(offset: usize, buf_len: usize) -> bool {
+    if offset % BLOCK_SIZE == 0 && buf_len > 0 && buf_len % BLOCK_SIZE == 0 {
+        true
+    } else {
+        false
+    }
+}
+
+/// Metadata for SwornDisk.
+#[derive(Debug)]
+pub struct SwornDiskMeta {
+    size: usize,
+    root_key: AeadKey,
+    image_dir: PathBuf,
+    is_setup: bool,
+}
+
+impl Default for SwornDiskMeta {
+    fn default() -> Self {
+        Self {
+            size: 0,
+            root_key: AeadKey::default(),
+            image_dir: PathBuf::from("run"),
+            is_setup: false,
+        }
+    }
+}
+
+impl SwornDiskMeta {
+    pub fn setup(
```

Comment:
If you mount multiple ext2 filesystems, I think `SwornDiskMeta` should not just be a global static variable.
`SwornDiskMeta::setup(.....) -> Result<SwornDiskMeta>`

Reply:
Yes, since currently we only have one device mount point "/dev/sworndisk" to serve only one ext2, so I force SwornDisk and SwornDiskMeta to be singletoned. In the future we can add multiple "/dev/sworndisk*" (multiple SwornDisk instance) to the DevFS to serve multiple Ext2s

Reply:
I agree